### PR TITLE
Add logging to capture swap requests, partner offers, and errors.

### PR DESCRIPTION
**Pull request by peachbits**

### CODE REVIEW

File: src/swap/foxExchange.js [hunk 1]
```diff
@@ -179,26 +196,33 @@ export function makeFoxExchangePlugin(
...
-              await request.fromWallet.denominationToNative(
-                String(rateResp.limitMaxDepositCoin),
-                request.fromCurrencyCode
-              )
+            const nativeMax = await request.fromWallet.denominationToNative(
+              String(rateResp.limitMaxDepositCoin),
+              request.fromCurrencyCode
             )
+            log.warn(
+              `${pluginId} SwapBelowLimitError\n${JSON.stringify(
```
Comment (paullinator) [inline src/swap/foxExchange.js:219]: should be SwapAboveLimitError

File: src/swap/changelly.js [hunk 2]
```diff
@@ -264,7 +268,7 @@ export function makeChangellyPlugin(
         },
         promoCode
       )
-      checkReply(sendReply, request)
+      checkReply(sendReply, request, log)
```
Comment (paullinator) [inline src/swap/changelly.js:271]: Line 193 should have a log noting that the fixed rate quote failed. As well, it should log if the estimate quote also fails

File: src/swap/coinswitch.js [hunk 3]
```diff
@@ -194,10 +197,20 @@ export function makeCoinSwitchPlugin(
...
+          `${pluginId} SwapBelowLimitError\n${JSON.stringify(
+            swapInfo
+          )}\n${nativeMin}`
+        )
         throw new SwapBelowLimitError(swapInfo, nativeMin)
       }
 
       if (gt(fromNativeAmount, nativeMax)) {
+        log.warn(
+          `${pluginId} SwapBelowLimitError\n${JSON.stringify(
```
Comment (paullinator) [inline src/swap/coinswitch.js:210]: SwapAboveLimitError

File: src/swap/coinswitch.js [hunk 4]
```diff
@@ -338,10 +353,20 @@ export function makeCoinSwitchPlugin(
...
```
Comment (paullinator) [inline src/swap/coinswitch.js:366]: SwapAboveLimitError

File: src/swap/faast.js [hunk 5]
```diff
@@ -101,6 +101,11 @@ export function makeFaastPlugin(opts: EdgeCorePluginOptions): EdgeSwapPlugin {
       replyJson != null &&
       /geo/.test(replyJson.error)
     ) {
+      log.warn(
+        `${pluginId} SwapBelowLimitError\n${JSON.stringify(
```
Comment (paullinator) [inline src/swap/faast.js:105]: SwapPermissionError

File: src/swap/faast.js [hunk 6]
```diff
@@ -189,6 +198,11 @@ export function makeFaastPlugin(opts: EdgeCorePluginOptions): EdgeSwapPlugin {
         (geoInfo.restricted &&
           (fromCurrency.restricted || toCurrency.restricted))
       ) {
+        log.warn(
+          `${pluginId} SwapBelowLimitError\n${JSON.stringify(
```
Comment (paullinator) [inline src/swap/faast.js:202]: SwapPermissionError

File: src/swap/faast.js [hunk 7]
```diff
@@ -281,9 +297,19 @@ export function makeFaastPlugin(opts: EdgeCorePluginOptions): EdgeSwapPlugin {
...
+            swapInfo
+          )}\n${nativeMin}`
+        )
         throw new SwapBelowLimitError(swapInfo, nativeMin)
       }
       if (nativeMax != null && gt(nativeAmount, nativeMax)) {
+        log.warn(
+          `${pluginId} SwapBelowLimitError\n${JSON.stringify(
```
Comment (paullinator) [inline src/swap/faast.js:309]: SwapAboveLimitError

File: src/swap/faast.js [hunk 8]
```diff
@@ -303,9 +329,19 @@ export function makeFaastPlugin(opts: EdgeCorePluginOptions): EdgeSwapPlugin {
...
+          log.warn(
+            `${pluginId} SwapBelowLimitError\n${JSON.stringify(swapInfo)}\n${
+              nativeMin || nativeAmount
+            }`
+          )
           throw new SwapBelowLimitError(swapInfo, nativeMin || nativeAmount)
         }
         if (/is greater/.test(e.message)) {
+          log.warn(
+            `${pluginId} SwapBelowLimitError\n${JSON.stringify(swapInfo)}\n${
```
Comment (paullinator) [inline src/swap/faast.js:341]: SwapAboveLimitError

File: src/swap/foxExchange.js [hunk 9]
```diff
@@ -119,20 +120,36 @@ export function makeFoxExchangePlugin(
...
           ) {
+            log.warn(`${pluginId} SwapCurrencyError ${JSON.stringify(request)}`)
             throw new SwapCurrencyError(
               swapInfo,
               request.fromCurrencyCode,
               request.toCurrencyCode
             )
           } else if (json.code === 'region_lock') {
+            log.warn(
+              `${pluginId} SwapBelowLimitError\n${JSON.stringify(
```
Comment (paullinator) [inline src/swap/foxExchange.js:131]: SwapPermissionError

File: src/swap/foxExchange.js [hunk 10]
```diff
@@ -119,20 +120,36 @@ export function makeFoxExchangePlugin(
...
+              `${pluginId} SwapBelowLimitError\n${JSON.stringify(swapInfo)}\n${
+                request.nativeAmount
+              }`
+            )
             throw new SwapBelowLimitError(swapInfo, request.nativeAmount)
           } else if (json.code === 'max_limit_breached') {
             // TODO: Using the nativeAmount here is technically a bug,
             // since we don't know the actual limit in this case:
+            log.warn(
+              `${pluginId} SwapBelowLimitError\n${JSON.stringify(swapInfo)}\n${
```
Comment (paullinator) [inline src/swap/foxExchange.js:149]: SwapAboveLimitError

**Review by paullinator — CHANGES_REQUESTED** [review]
Please check that the errors logged are appropriate for the actual error. I marked a dozen but I did not fully check the rest of the PR. It looks like it is a recurring issue.

File: src/swap/changelly.js [hunk 11]
```diff
@@ -169,11 +183,13 @@ export function makeChangellyPlugin(
       userSettings: Object | void,
       opts: { promoCode?: string }
     ): Promise<EdgeSwapQuote> {
+      log.warn(`${pluginId} swap requested ${JSON.stringify(request)}`)
```
Comment (swansontec) [inline src/swap/changelly.js:186]: Again, we are going to get a ton of redundant logs this way. The core should simply log this exactly once before even calling the plugins.

File: src/swap/changelly.js [hunk 12]
```diff
@@ -114,18 +116,30 @@ async function getAddress(
...
 }
 
-function checkReply(reply: Object, request: EdgeSwapRequest) {
+function checkReply(reply: Object, request: EdgeSwapRequest, log: EdgeLog) {
   if (reply.error != null) {
     if (
       reply.error.code === -32602 ||
       /Invalid currency:/.test(reply.error.message)
     ) {
+      log.warn(`${pluginId} SwapCurrencyError ${JSON.stringify(request)}`)
```
Comment (swansontec) [inline src/swap/changelly.js:125]: We should probably not log these errors - the core already has centralized logging logic for dealing with all the swaps as a group. If we honestly want all errors logged, we should simply adjust the core, not all the plugins.

**Review by swansontec — COMMENTED** [review]
I agree that we should return more specific errors (above / below limits, for instance), but logging the inputs & outputs (or errors) of every single quote plugin is not a good approach - the core can simply log the swap request, and then log all the results (or errors) as they come in. We already have similar logic in the core, which tried to be a bit more balanced in its reporting, but if we want to turn up the verbosity, let's do it in the core.

**Comment by paullinator** [comment]
Per @swansontec's suggestions, it looks like logging is best done in the core and this pr can be closed.

**Pull request closed without merging**